File: 417.pacific-atlantic-water-flow.java
/*
 * @lc app=leetcode id=417 lang=java
 *
 * [417] Pacific Atlantic Water Flow
 */

// @lc code=start
class Solution {

    static final int[][] moves = new int[][] {{-1,0},{0,1},{1,0},{0,-1}};
    List<List<Integer>> res = new LinkedList<>();

    public List<List<Integer>> pacificAtlantic(int[][] matrix) {
        if (matrix == null || matrix.length == 0 || matrix[0].length == 0) return res;

        boolean[][] pacificReachable = new boolean[matrix.length][matrix[0].length];
        boolean[][] atlanticReachable = new boolean[matrix.length][matrix[0].length];

        for (int i = 0; i < matrix.length; ++i) {
            dfs(matrix, pacificReachable, i, 0);
            dfs(matrix, atlanticReachable, i, matrix[0].length-1);
        }
        for (int j = 0; j < matrix[0].length; ++j) {
            dfs(matrix, pacificReachable, 0, j);
            dfs(matrix, atlanticReachable, matrix.length-1, j);
        }
        for (int i = 0; i < matrix.length; ++i) {
            for (int j = 0; j < matrix[0].length; ++j) {
                if (pacificReachable[i][j] && atlanticReachable[i][j]) {
                    res.add(Arrays.asList(new Integer[]{i, j}));
                }
            }
        }
        return res;
    }

    private void dfs(int[][] matrix, boolean[][] reachable, int i, int j) {
        dfs(matrix, reachable, i, j, matrix[i][j]);
    }

    private void dfs(int[][] matrix, boolean[][] reachable, int i, int j, int toVal) {
        if (matrix[i][j] < toVal || reachable[i][j]) return;
        reachable[i][j] = true;
        for (int[] move : moves) {
            int ni = i + move[0], nj = j + move[1];
            if (ni < 0 || ni >= matrix.length || nj < 0 || nj >= matrix[0].length) {
                continue;
            }
            dfs(matrix, reachable, ni, nj, matrix[i][j]);
        }
    }

}
// @lc code=end

File: go/417.pacific-atlantic-water-flow.go
/*
 * @lc app=leetcode id=417 lang=golang
 *
 * [417] Pacific Atlantic Water Flow
 */

// @lc code=start


var neigborDist = [][]int{[]int{-1,0},[]int{1,0},[]int{0,-1},[]int{0,1}}

func pacificAtlantic(matrix [][]int) [][]int {
	res := [][]int{}
	if len(matrix) == 0 || len(matrix[0]) == 0 {
		return res
	}
	toPacific := make([][]bool, len(matrix))
	toAtlantic := make([][]bool, len(matrix))
	for i := 0; i < len(matrix); i++ {
		toPacific[i] = make([]bool, len(matrix[0]))
		toAtlantic[i] = make([]bool, len(matrix[0]))
	}
	for i := 0; i < len(matrix); i++ {
		dfs(i, 0, matrix[i][0], &matrix, &toPacific)
		dfs(i, len(matrix[0])-1, matrix[i][len(matrix[0])-1], &matrix, &toAtlantic)
	}
	for j := 0; j < len(matrix[0]); j++ {
		dfs(0, j, matrix[0][j], &matrix, &toPacific)
		dfs(len(matrix)-1, j, matrix[len(matrix)-1][j], &matrix, &toAtlantic)
	}

	for i := 0; i < len(matrix); i++ {
		for j := 0; j < len(matrix[0]); j++ {
			if toPacific[i][j] && toAtlantic[i][j] {
				res = append(res, []int{i, j})
			}
		}
	}
	return res
}

func dfs(i, j, prev int, mtx *[][]int, toboundary *[][]bool) {
	matrix := *mtx
	if i < 0 || i >= len(matrix) {
		return
	}
	if j < 0 || j >= len(matrix[0]) {
		return
	}
	if matrix[i][j] < prev || (*toboundary)[i][j] {
		return
	}
	(*toboundary)[i][j] = true
	dfs(i-1, j, matrix[i][j], mtx, toboundary)
	dfs(i+1, j, matrix[i][j], mtx, toboundary)
	dfs(i, j-1, matrix[i][j], mtx, toboundary)
	dfs(i, j+1, matrix[i][j], mtx, toboundary)
}